nano_graphrag/_storage/gdb_nebula.py

```diff
@@ -0,0 +1,974 @@
+                if not result.is_succeeded():
+                    raise RuntimeError(f"update {node_id} clusters failed: {result.error_msg()}")
+
+        update_tasks = []
+        for node_id, clusters in cluster_data.items():
+            update_tasks.append(update_node_clusters(node_id, clusters))
+
+        await asyncio.gather(*update_tasks)
+
+        # Community Vertex data
+        ng_nx_community_writer = self.writer_cls(
+            data=data,
+            nebula_config=self.config,
+        )
+        ng_nx_community_writer.set_options(
+            label=self.COMMUNITY_VERTEX_TYPE,
+            properties=["level", "cluster"],
+            write_mode="insert",
+            sink="nebulagraph_vertex",
+        )
+        ng_nx_community_writer.write()
+
+        ng_nx_community_edge_writer = self.writer_cls(
+            data=edges,
+            nebula_config=self.config,
+        )
+        ng_nx_community_edge_writer.set_options(
+            label=self.COMMUNITY_EDGE_TYPE,
+            properties=["level"],
+            write_mode="insert",
+            sink="nebulagraph_edge",
+        )
+        ng_nx_community_edge_writer.write()
+
+        return
+
+
+    async def _leiden_clustering(self):
+        from graspologic.partition import hierarchical_leiden
+
+        # TODO: introduce Cache mechanism for this.
+        self._graph = self.reader.read()
+        self._graph_homogenous = nx.Graph(self._graph)
+
+        graph = NetworkXStorage.stable_largest_connected_component(self._graph_homogenous)
+        community_mapping = hierarchical_leiden(
+            graph,
+            max_cluster_size=self.global_config["max_graph_cluster_size"],
+            random_seed=self.global_config["graph_cluster_seed"],
+        )
+
+        node_communities: dict[str, list[dict[str, str]]] = defaultdict(list)
+        __levels = defaultdict(set)
+
+        for partition in community_mapping:
+            level_key = partition.level
+            cluster_id = partition.cluster
+            node_communities[partition.node].append(
+                {"level": level_key, "cluster": cluster_id}
+            )
+            __levels[level_key].add(cluster_id)
+        node_communities = dict(node_communities)
+        __levels = {k: len(v) for k, v in __levels.items()}
+        logger.info(f"Each level has communities: {dict(__levels)}")
+        await self._cluster_data_to_graph(node_communities)
+
+    async def community_schema(self) -> dict[str, SingleCommunitySchema]:
+        results = defaultdict(
+            lambda: dict(
+                level=None,
+                title=None,
+                edges=set(),
+                nodes=set(),
+                chunk_ids=set(),
+                occurrence=0.0,
+                sub_communities=[],
+            )
+        )
+        max_num_ids = 0
+        levels = defaultdict(set)
+
+
+        communities_result = self.client.execute_py(
```

Comment:
this involves fullscan, we could create a index for COMMUNITY_VERTEX_TYPE tag to make this call more scalable :)

Ref: https://docs.nebula-graph.io/3.8.0/3.ngql-guide/14.native-index-statements/1.create-native-index/

If we are going do it in a programming way, a pitfall is that the index need to wait for two heartbeats of NebulaGraph cluster(default 20s) maximally to ensure it's synced across all metaDs

We could keep it as is or ask user to create them manually.

Note if we create index after the data was already there, we need to rebuild the index(see from the docs as well)

Reply:
I want to know why community_report_storage_cls still use JsonKVStorage ? Is there any problem

Reply:
It shouldn't  actually our preparatory impl are full-graph-wise. Just to kept the original design of ms and nano graph-rag here.